File: src/types/dicom.ts
/**
 * DICOM TypeScript type definitions
 */

/**
 * DICOM modality types
 */
export type Modality = 'CT' | 'MR' | 'XR' | 'US' | 'NM' | 'PT' | 'CR' | 'DX' | 'MG' | 'OT';

/**
 * DICOM Study from API
 */
export interface Study {
  id: string;
  studyInstanceUID: string;
  studyDate: string | null;
  studyTime: string | null;
  studyDescription: string | null;
  accessionNumber: string | null;
  patientId: string | null;
  patientName: string | null;
  patientBirthDate: string | null;
  patientSex: string | null;
  modality: Modality;
  numberOfSeries: number;
  numberOfInstances: number;
  createdAt: string;
  updatedAt: string;
}

/**
 * DICOM Series from API
 */
export interface Series {
  id: string;
  seriesInstanceUID: string;
  studyInstanceUID: string;
  seriesNumber: number | null;
  seriesDescription: string | null;
  modality: Modality;
  bodyPart: string | null;
  numberOfInstances: number;
  createdAt: string;
}

/**
 * DICOM Instance from API
 */
export interface Instance {
  id: string;
  sopInstanceUID: string;
  seriesInstanceUID: string;
  instanceNumber: number | null;
  rows: number;
  columns: number;
  bitsAllocated: number;
  bitsStored: number;
  pixelRepresentation: number;
  windowCenter: number | null;
  windowWidth: number | null;
  rescaleIntercept: number | null;
  rescaleSlope: number | null;
  sliceThickness: number | null;
  sliceLocation: number | null;
  imagePositionPatient: string | null;
  imageOrientationPatient: string | null;
  pixelSpacing: string | null;
  createdAt: string;
}

/**
 * Instance with optional URL properties (used during loading)
 */
export interface InstanceWithUrls extends Instance {
  _staticUrl?: string;
  _localUrl?: string;
}

/**
 * Study with series included
 */
export interface StudyWithSeries extends Study {
  series: SeriesWithInstances[];
}

/**
 * Series with instances included
 */
export interface SeriesWithInstances extends Series {
  instances: Instance[];
}

/**
 * Series with instances that may have URLs
 */
export interface SeriesWithInstanceUrls extends Series {
  instances: InstanceWithUrls[];
}

/**
 * Study with series that may have URL instances
 */
export interface StudyWithSeriesUrls extends Study {
  series: SeriesWithInstanceUrls[];
}

/**
 * Study list filters
 */
export interface StudyFilters {
  patientName?: string;
  patientId?: string;
  modality?: Modality;
  studyDateFrom?: string;
  studyDateTo?: string;
  accessionNumber?: string;
  page?: number;
  limit?: number;
}

/**
 * Paginated response
 */
export interface PaginatedResponse<T> {
  items: T[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

/**
 * DICOM upload request
 */
export interface DicomUploadRequest {
  files: File[];
  projectId?: string;
}

/**
 * DICOM upload response
 */
export interface DicomUploadResponse {
  studyInstanceUID: string;
  numberOfFilesProcessed: number;
  numberOfSeriesCreated: number;
  numberOfInstancesCreated: number;
}

/**
 * WADO-RS metadata request
 */
export interface WadoMetadataRequest {
  studyInstanceUID: string;
  seriesInstanceUID?: string;
  sopInstanceUID?: string;
}

/**
 * Viewport display state
 */
export interface ViewportDisplayState {
  windowCenter: number;
  windowWidth: number;
  zoom: number;
  pan: { x: number; y: number };
  rotation: number;
  flipHorizontal: boolean;
  flipVertical: boolean;
  invert: boolean;
}

/**
 * DICOM Redux state
 */
export interface DicomState {
  // Currently loaded study/series
  currentStudy: Study | null;
  currentSeries: Series | null;
  currentInstances: Instance[];

  // Studies list
  studies: Study[];
  studiesLoading: boolean;
  studiesError: string | null;

  // Pagination
  totalStudies: number;
  currentPage: number;
  pageSize: number;

  // Filters
  filters: StudyFilters;

  // Viewport state
  activeViewportId: string | null;
  viewportStates: Record<string, ViewportDisplayState>;

  // Image loading
  imageLoadingProgress: number;
  isLoadingImages: boolean;
}

/**
 * Initial DICOM state
 */
export const initialDicomState: DicomState = {
  currentStudy: null,
  currentSeries: null,
  currentInstances: [],
  studies: [],
  studiesLoading: false,
  studiesError: null,
  totalStudies: 0,
  currentPage: 1,
  pageSize: 20,
  filters: {},
  activeViewportId: null,
  viewportStates: {},
  imageLoadingProgress: 0,
  isLoadingImages: false,
};

/**
 * Extended study type that may include series
 */
export interface StudyMayHaveSeries extends Study {
  series?: SeriesWithInstances[];
}

/**
 * Study filter key type for type-safe filter updates
 */
export type StudyFilterKey = keyof StudyFilters;

/**
 * Cornerstone Window Extension
 * Extends Window interface to include Cornerstone global references
 */
declare global {
  interface Window {
    __cornerstoneCore?: typeof import('@cornerstonejs/core');
    __cornerstoneWadouriFileManager?: {
      add: (file: File) => string;
      get: (id: string) => File | undefined;
      remove: (id: string) => void;
      purge: () => void;
    };
  }
}

/**
 * Cornerstone Viewport with optional image property
 */
export interface CornerstoneViewportWithImage {
  csImage?: {
    imageId: string;
    minPixelValue?: number;
    maxPixelValue?: number;
    windowCenter?: number;
    windowWidth?: number;
    [key: string]: unknown;
  };
  getProperties?: () => {
    voiRange?: { lower: number; upper: number };
    [key: string]: unknown;
  };
}

/**
 * Cornerstone Annotation Event
 */
export interface CornerstoneAnnotationEvent {
  detail?: {
    annotation?: {
      annotationUID: string;
      data: {
        handles?: {
          points?: Array<{ x: number; y: number; z?: number }>;
          textBox?: { worldPosition: number[] };
        };
        cachedStats?: Record<string, unknown>;
        text?: string;
        label?: string;
      };
      metadata?: {
        toolName?: string;
        viewportId?: string;
        referencedImageId?: string;
      };
      isVisible?: boolean;
      isLocked?: boolean;
    };
    viewportId?: string;
    element?: HTMLElement;
  };
}

/**
 * Cornerstone Tools module type
 */
export interface CornerstoneToolsModule {
  annotation: {
    state: {
      getAnnotations: (
        toolName: string,
        element?: HTMLElement
      ) => Array<{
        annotationUID: string;
        data: Record<string, unknown>;
        metadata?: Record<string, unknown>;
      }>;
      addAnnotation: (annotation: unknown, element?: HTMLElement) => void;
      removeAnnotation: (annotationUID: string) => void;
    };
  };
  Enums: {
    Events: {
      ANNOTATION_ADDED: string;
      ANNOTATION_MODIFIED: string;
      ANNOTATION_REMOVED: string;
      ANNOTATION_COMPLETED: string;
    };
    [key: string]: unknown;
  };
  [key: string]: unknown;
}

File: services/dicom-service/app/schemas/dicom.py
"""
Pydantic schemas for DICOM Service
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime, date


# Study Schemas
class StudyResponse(BaseModel):
    """Study response"""
    id: str
    study_instance_uid: str
    orthanc_id: Optional[str]
    patient_id: Optional[str]
    patient_name: Optional[str]
    patient_birth_date: Optional[date]
    patient_sex: Optional[str]
    patient_age: Optional[str]
    study_date: Optional[date]
    study_time: Optional[str]
    study_description: Optional[str]
    accession_number: Optional[str]
    institution_name: Optional[str]
    referring_physician_name: Optional[str]
    modalities: Optional[List[str]]
    number_of_series: int
    number_of_instances: int
    total_size: int
    is_processed: bool
    is_anonymized: bool
    uploaded_by: Optional[str]
    created_at: datetime
    updated_at: datetime


class StudyListRequest(BaseModel):
    """Study list request"""
    patient_id: Optional[str] = None
    patient_name: Optional[str] = None
    study_date_from: Optional[date] = None
    study_date_to: Optional[date] = None
    modality: Optional[str] = None
    accession_number: Optional[str] = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)


class StudyListResponse(BaseModel):
    """Study list response"""
    studies: List[StudyResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


# Series Schemas
class SeriesResponse(BaseModel):
    """Series response"""
    id: str
    series_instance_uid: str
    orthanc_id: Optional[str]
    study_id: str
    series_number: Optional[int]
    series_description: Optional[str]
    modality: Optional[str]
    body_part_examined: Optional[str]
    protocol_name: Optional[str]
    number_of_instances: int
    rows: Optional[int]
    columns: Optional[int]
    slice_thickness: Optional[str]
    pixel_spacing: Optional[str]
    total_size: int
    created_at: datetime
    updated_at: datetime


class SeriesListResponse(BaseModel):
    """Series list response"""
    series: List[SeriesResponse]
    total: int


# Instance Schemas
class InstanceResponse(BaseModel):
    """Instance response"""
    id: str
    sop_instance_uid: str
    orthanc_id: Optional[str]
    series_id: str
    instance_number: Optional[int]
    sop_class_uid: Optional[str]
    rows: Optional[int]
    columns: Optional[int]
    bits_allocated: Optional[int]
    bits_stored: Optional[int]
    image_position_patient: Optional[str]
    image_orientation_patient: Optional[str]
    slice_location: Optional[str]
    file_size: int
    created_at: datetime
    updated_at: datetime


class InstanceListResponse(BaseModel):
    """Instance list response"""
    instances: List[InstanceResponse]
    total: int


# Upload Schemas
class DicomUploadResponse(BaseModel):
    """DICOM upload response"""
    success: bool
    study_instance_uid: str
    series_instance_uids: List[str]
    instance_count: int
    orthanc_id: Optional[str]
    message: str


# Statistics Schemas
class DicomStatisticsResponse(BaseModel):
    """DICOM statistics response"""
    total_studies: int
    total_series: int
    total_instances: int
    total_size: int
    studies_by_modality: Dict[str, int]
    storage_by_modality: Dict[str, int]


# Search Schemas
class DicomSearchRequest(BaseModel):
    """DICOM search request"""
    query: Optional[str] = None
    patient_id: Optional[str] = None
    patient_name: Optional[str] = None
    study_date_from: Optional[date] = None
    study_date_to: Optional[date] = None
    modality: Optional[str] = None
    body_part: Optional[str] = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)


# Tag Schemas
class DicomTagsResponse(BaseModel):
    """DICOM tags response"""
    study_instance_uid: str
    tags: Dict[str, Any]


# Preview Schemas
class InstancePreviewRequest(BaseModel):
    """Instance preview request"""
    instance_id: str
    quality: int = Field(default=90, ge=1, le=100)


# Anonymization Schemas
class AnonymizeRequest(BaseModel):
    """Anonymization request"""
    study_id: str
    keep_patient_id: bool = True
    keep_study_date: bool = True
    new_patient_id: Optional[str] = None
    new_patient_name: Optional[str] = None


class AnonymizeResponse(BaseModel):
    """Anonymization response"""
    success: bool
    original_study_id: str
    new_study_id: Optional[str]
    message: str

class BatchAnonymizeRequest(BaseModel):
    """Batch anonymization request"""
    study_ids: List[str]
    keep_patient_id: bool = False
    keep_study_date: bool = False
    keep_descriptive_tags: bool = True
    new_patient_prefix: Optional[str] = "ANON"
    salt: Optional[str] = None